src/hooks/useUploadsManager/useStartFileUpload.tsx

```diff
@@ -0,0 +1,207 @@
+          setAssetUploadProgress((loaded / total) * 100)
+        }
+
+        pendingUploadingNotificationsCounts.current++
+        displayUploadingNotification.current()
+
+        assetsNotificationsCount.current.uploads[assetKey] =
+          (assetsNotificationsCount.current.uploads[assetKey] || 0) + 1
+
+        await axios.put(assetUrl.toString(), opts?.changeHost ? fileInState?.blob : file, {
+          headers: {
+            // workaround for a bug in the storage node
+            'Content-Type': '',
+          },
+          raxConfig: {
+            retry: RETRIES_COUNT,
+            noResponseRetries: RETRIES_COUNT,
+            // add 400 to default list of codes to retry
+            // seems storage node sometimes fails to calculate the IFPS hash correctly
+            // trying again in that case should succeed
+            statusCodesToRetry: [
+              [100, 199],
+              [400, 400],
+              [429, 429],
+              [500, 599],
+            ],
+            retryDelay: RETRY_DELAY,
+            backoffType: 'static',
+            onRetryAttempt: (err) => {
+              const cfg = rax.getConfig(err)
+              if (cfg?.currentRetryAttempt === 1) {
+                updateAsset(asset.contentId, 'reconnecting')
+              }
+            },
+          },
+          onUploadProgress: setUploadProgress,
+        })
+
+        // Cancel delayed functions that would overwrite asset status back to 'inProgres'
+
+        // TODO: remove assets from the same parent if all finished
+        updateAsset(asset.contentId, 'completed')
+        setAssetUploadProgress(100)
+        assetsNotificationsCount.current.uploaded[assetKey] =
+          (assetsNotificationsCount.current.uploaded[assetKey] || 0) + 1
+        displayUploadedNotification.current(assetKey)
+      } catch (e) {
+        Logger.error('Failed to upload to storage provider', { storageUrl, error: e })
+        updateAsset(asset.contentId, 'error')
+        setAssetUploadProgress(0)
+
+        const axiosError = e as AxiosError
+        const networkFailure =
+          axiosError.isAxiosError &&
+          (!axiosError.response?.status || (axiosError.response.status < 400 && axiosError.response.status >= 500))
+        if (networkFailure) {
+          markStorageProviderNotWorking(storageProviderId)
+        }
+
+        const snackbarDescription = networkFailure ? 'Host is not responding' : 'Unexpected error occurred'
+        displaySnackbar({
+          title: 'Failed to upload asset',
+          description: snackbarDescription,
+          actionText: 'Go to uploads',
+          onActionClick: () => navigate(absoluteRoutes.studio.uploads()),
+          iconType: 'warning',
+        })
+      }
+    },
+    [
+      addAsset,
+      assetsFiles,
+      displaySnackbar,
+      getStorageProvider,
+      markStorageProviderNotWorking,
+      navigate,
+      setAssetsFiles,
+      setUploadsProgress,
+      updateAsset,
+    ]
+  )
```

Comment:
Tying to understand if it would it be possible to have this logic inside the store or are we limited by something here

Reply:
It actually was part of the store previously but I asked @drillprop to move it here. The reason is I'd like to separate store (state) and logic for it. I'm aware this may not be the most usual approach with libraries like Redux where you would put all your HTTP calls in the store, but I'd like our store to be lightweight, let's try it out